docs/enhancements/api-for-skipping-validations.md

```diff
@@ -0,0 +1,89 @@
+---
+title: api-for-skipping-validations
+authors:
+  - "@pmaidmen"
+creation-date: 2023-01-08
+last-updated: 2023-01-08
+---
+
+# Allow validations to be skipped
+
+# Summary
+
+We need a new api that will allow us to skip cluster/host validations, this will be for unsupported use.
+This API should work on a per cluster basis and will have it's own feature flag to allow us to enable it for specific users.
+An event should be emitted whenever the API is used during an installation.
+Clusters containing skipped validations should not contribute to statistics as these may be skewed.
+# Motivation
+
+Some users have very specific HW that doesn't pass our validations and we want to allow them to proceed with installation.
+Sometimes we have bugs in our validations that block people from installing and we don't want our partners left in a position where they can't install.
+
+### Goals
+
+- Allow validations to be bypassed so that installation may proceed.
+
+### Non-Goals
+
+- None identified at this stage
+
+### The implementation of skipped-validations
+
+The way we propose to deal with the validation is as follows
+
+#### Propoal - Run all validations, allow the user to review but proceed anyway. (Suggested by Michael Hrivnak as an intepretation of requirements)
+
+    * All validations run.
+    * One or more validations fail and describe a problem.
+    * The user evaluates the alleged problems and decides to try the installation anyway.
+    * When skipping validations, there will be some "absolutely mandatory" validations that cannot fail prior to proceeding.
+        - This list is TBC
+    * If the user skips validations (by making a change to a boolean in the cluster object), we should
```

Comment:
Instead of a boolean, let's have a string with a list of validations to skip, and maybe a special "all" value.

Reply:
This seems at odds with this recommendation. 

"Again, seems overly complex - we can just add an annotation to skip all validations"

and 

"This seems very complex, both in the API and in the DB. Why not just:
PUT api/v2/cluster//skip-host-validations (body has true/false)
GET api/v2/cluster//skip-host-validations (returns true/false)"

Both of which seem to imply that we want a simple on/off switch for ALL validations, I think I may have misunderstood?

Can you clarify further  on what is actually needed here?

Reply:
I think that there are different opinions here also with regards to the requirements.
In my view, we should:
1. Allow to ignore all validations
2. API only - no impact whatsoever to the UI
3. We can have this gates by AMS capabilities - so it will be used only for whom we allow it (in my view we can also just open it to everyone, but if someone is concerned people will abuse it, we can allow it only to specific organizations)
4. We should put a flag on clusters that the user installed while ignoring validations.

Reply:
I was previously put off by the number of APIs to manage the list of validations to skip (there were operations to add, remove, list, ...).  However, boolean APIs are fragile because you need to break the API to add a third option.  So now I am suggesting a compromise, where we still only have one PUT and one GET API, but instead of having a boolean in the body, we have a list of strings.

I agree with Oved's requirements above, just in addition to allowing us to ignore all validations, we can enable ignoring some validations as well.

Reply:
OK, agreed that we will allow a string input with a list of strings for a single PUT and a single GET API, I'll update the document now.

Reply:
That validation is only looking at the value in the DB - the actual measurement happens as part of the preparing-to-install phase.  Are you suggesting that we skip the actual measurement?  That's a very different behavior.  Have we seen other cases where the actual measurement was problematic?  I know that the connectivity check failed in a few cases.

Reply:
Yeah, I'm referring to the actual measurement that could take a while. I've seen an issue with that once where it timeouts after 10mins (using a VM though).